File: cli/src/botc_gender/formats/app_schema.py
from __future__ import annotations

from typing import Any

from botc_gender.formats.strategies.official_override import resolve_role_id
from botc_gender.formats.strategies.types import Strategy
from botc_gender.images import official_image_urls


def build_character_object(
    role_id: str,
    texts: dict[str, Any],
    metadata: dict[str, Any],
    *,
    strategy: Strategy,
    suffix: str = "_de",
    use_official_images: bool = False,
) -> dict[str, Any]:
    output_id = resolve_role_id(role_id, strategy, suffix)

    character: dict[str, Any] = {
        "id": output_id,
        "name": str(texts["name"])[:30],
        "team": metadata["team"],
        "ability": str(texts["ability"])[:250],
    }

    if metadata.get("edition"):
        character["edition"] = metadata["edition"]
    if use_official_images:
        character["image"] = official_image_urls(
            role_id=role_id,
            edition=str(metadata.get("edition", "")),
            team=metadata["team"],
        )
    elif metadata.get("image"):
        character["image"] = metadata["image"]
    if metadata.get("firstNight") is not None:
        character["firstNight"] = metadata["firstNight"]
    if metadata.get("otherNight") is not None:
        character["otherNight"] = metadata["otherNight"]
    if metadata.get("setup") is not None:
        character["setup"] = metadata["setup"]
    if metadata.get("special"):
        character["special"] = metadata["special"]

    if texts.get("firstNightReminder"):
        character["firstNightReminder"] = str(texts["firstNightReminder"])[:500]
    if texts.get("otherNightReminder"):
        character["otherNightReminder"] = str(texts["otherNightReminder"])[:500]
    if texts.get("flavor"):
        character["flavor"] = str(texts["flavor"])[:500]
    if texts.get("reminders"):
        character["reminders"] = [str(item)[:30] for item in texts["reminders"]][:20]
    if metadata.get("remindersGlobal"):
        character["remindersGlobal"] = metadata["remindersGlobal"]

    return character


def build_script_array(
    meta: dict[str, Any],
    characters: list[dict[str, Any]],
) -> list[Any]:
    return [meta, *characters]

File: web/src/lib/appSchema.ts
import type { Strategy } from "./strategies";
import { resolveRoleId } from "./strategies";
import { officialImageUrls } from "./images";

export interface RoleTexts {
  name?: string;
  ability?: string;
  firstNightReminder?: string;
  otherNightReminder?: string;
  flavor?: string;
  reminders?: string[];
}

export function buildCharacterObject(
  roleId: string,
  texts: RoleTexts,
  metadata: Record<string, unknown>,
  options: {
    strategy: Strategy;
    suffix?: string;
    useOfficialImages?: boolean;
  },
): Record<string, unknown> {
  const outputId = resolveRoleId(roleId, options.strategy, options.suffix);
  const character: Record<string, unknown> = {
    id: outputId,
    name: String(texts.name ?? "").slice(0, 30),
    team: metadata.team,
    ability: String(texts.ability ?? "").slice(0, 250),
  };

  if (metadata.edition) {
    character.edition = metadata.edition;
  }
  if (options.useOfficialImages) {
    character.image = officialImageUrls(
      roleId,
      String(metadata.edition ?? ""),
      String(metadata.team ?? ""),
    );
  } else if (metadata.image) {
    character.image = metadata.image;
  }
  if (metadata.firstNight !== undefined && metadata.firstNight !== null) {
    character.firstNight = metadata.firstNight;
  }
  if (metadata.otherNight !== undefined && metadata.otherNight !== null) {
    character.otherNight = metadata.otherNight;
  }
  if (metadata.setup !== undefined && metadata.setup !== null) {
    character.setup = metadata.setup;
  }
  if (metadata.special) {
    character.special = metadata.special;
  }

  if (texts.firstNightReminder) {
    character.firstNightReminder = String(texts.firstNightReminder).slice(0, 500);
  }
  if (texts.otherNightReminder) {
    character.otherNightReminder = String(texts.otherNightReminder).slice(0, 500);
  }
  if (texts.flavor) {
    character.flavor = String(texts.flavor).slice(0, 500);
  }
  if (texts.reminders) {
    character.reminders = texts.reminders.map((item) => String(item).slice(0, 30)).slice(0, 20);
  }
  if (metadata.remindersGlobal) {
    character.remindersGlobal = metadata.remindersGlobal;
  }

  return character;
}

export function buildScriptArray(
  meta: Record<string, unknown>,
  characters: Record<string, unknown>[],
): unknown[] {
  return [meta, ...characters];
}
